### PR TITLE
Enhance GHA to check uncommitted changes

### LINKED ISSUE #16: GHA enhancement no uncommitted changes
continuation of https://github.com/opendatahub-io/model-registry/pull/6#pullrequestreview-1635832530

implement on GHA build like

```yaml
      - name: Check if there are uncommitted file changes
        run: |
          clean=$(git status --porcelain)
          if [[ -z "$clean" ]]; then
              echo "Empty git status --porcelain: $clean"
          else
              echo "Uncommitted file changes detected: $clean"
              git diff
              exit 1
          fi
```

to ensure all checked in generated files correspond to what can be generated by CI/CD.

### PR DESCRIPTION
**Pull request by tarilabs**
Resolves #16 

## How Has This Been Tested?
Will show results in Actions tab.
See for instance: https://github.com/opendatahub-io/model-registry/actions/runs/6430409857/job/17461329308?pr=63#step:9:1

## Merge criteria:

- [x] The commits and have meaningful messages; the author will squash them [after approval](https://github.com/opendatahub-io/opendatahub-community/blob/main/contributor-cheatsheet.md#:~:text=Usually%20this%20is%20done%20in%20last%20phase%20of%20a%20PR%20revision) or will ask to merge with squash.
- [n/a] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [n/a] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Review by lampajr — DISMISSED** [review]
lgtm

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -30,5 +30,15 @@ jobs:
         run: make clean
       - name: Build
         run: make build
+      - name: Check if there are uncommitted file changes
+        run: |
+          clean=$(git status --porcelain)
```
Comment (antisaling) [inline .github/workflows/build.yml:35]: When you do this you're transforming the newline-separated input into a string, so if there are multiple files they'll be echoed all in the same line (e.g. `M lib/src/git.rs ?? new_file`). I'd rather store this in an array, which you can accomplish using `mapfile`.

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -30,5 +30,15 @@ jobs:
...
+          if [[ -z "$clean" ]]; then
+              echo "Empty git status --porcelain: $clean"
```
Comment (antisaling) [inline .github/workflows/build.yml:37]: Also, no need to print anything here. You can simply test whether it's not empty using `-n`.

**Review by antisaling — COMMENTED** [review]
Left some suggestions as to how we could get a better-formatted output inline.

File: .github/workflows/build.yml (hunk 1 above)
Comment (tarilabs) [reply .github/workflows/build.yml:35]: > if there are multiple files they'll be echoed all in the same line

not sure I'm following; I have them echoed on multiple lines per the following trivial demo shows?

```
On branch tarilabs-20231006-gh50
Your branch is up to date with 'origin/tarilabs-20231006-gh50'.

Untracked files:
  (use "git add <file>..." to include in what will be committed)
	asd
	asd2

nothing added to commit but untracked files present (use "git add" to track)
model-registry %           clean=$(git status --porcelain)
          if [[ -z "$clean" ]]; then
              echo "Empty git status --porcelain: $clean"
          else
              echo "Uncommitted file changes detected: $clean"
              git diff
              exit 1
          fi
Uncommitted file changes detected: ?? asd
?? asd2
```

The `asd2` file is on separate line from file `asd`

Comment (antisaling) [reply .github/workflows/build.yml:35]: image

This is what I get running locally.

Comment (tarilabs) [reply .github/workflows/build.yml:35]: As this run demonstrates: https://github.com/opendatahub-io/model-registry/actions/runs/6469680763/job/17564213490?pr=63#step:11:20

image

output is displayed on multiple lines on GHA too.

Given this works good-enough and it's an embedded sh script in GHA, I would vote to proceed as-is, and the https://github.com/opendatahub-io/model-registry/pull/63#discussion_r1349514329 could be done in subsequent PRs, maybe externalizing the script if it's helpful to run it locally. 

wdyt?

File: .github/workflows/build.yml (hunk 2 above)
Comment (tarilabs) [reply .github/workflows/build.yml:37]:
```suggestion
          if [[ -n "$clean" ]]; then
              echo "Empty git status and no uncommitted files detected."
```

you mean like this?

Doesn't seem to work as intended? 🤔 

image

as is no longer detecting the uncommited changes.

**Comment by tarilabs** [comment]
no vetos, comments or missing.

@opendatahub-io/model-registry-maintainers I would need at least 1 approving review please to proceed to merge, please.

**Review by lampajr — APPROVED** [review]

File: .github/workflows/build.yml (hunk 2 above)
Comment (antisaling) [reply .github/workflows/build.yml:37]: I meant like so:

```bash
if [[ -n "$clean" ]]; then
    echo "Uncommitted file changes detected: $clean"
    git diff
    exit 1
fi
```

Sorry, I forgot to reply to this earlier.

**Review by antisaling — APPROVED** [review]
Approving as I don't want to stall you on this minor change.

**Pull request merged**